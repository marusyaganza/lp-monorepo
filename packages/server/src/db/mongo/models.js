const { formatIds, formatFilter } = require('../../utils/dbUtils');
const { ObjectId } = require('mongodb');

function createModel(db, table) {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result = await db.collection(table).find(formatFilter(filter));
    const data = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields) {
    const data = { ...fields };
    delete data.id;
    delete data.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    return result?.result;
  }

  async function deleteOne(filter) {
    console.log('filter', filter);
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    return result?.result;
  }

  async function createMany(entries) {
    const formattedEntries = entries.map(entry => ({
      ...entry,
      createdAt: Date.now()
    }));
    const result = await db.collection(table).insertMany(formattedEntries);
    return result.ops.map(formatIds);
  }

  return {
    findOne,
    findMany,
    createOne,
    createMany,
    updateOne,
    deleteOne
  };
}

module.exports = createModel;
