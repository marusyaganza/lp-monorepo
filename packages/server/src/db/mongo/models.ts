import { formatIds, formatFilter } from '../../utils/dbUtils';

import { Table, ModelType, DbData } from '..';

const { ObjectId } = require('mongodb');

function createModel<T extends DbData>(db: any, table: Table): ModelType<T> {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result: T[] = await db.collection(table).find(formatFilter(filter));
    const data: T[] = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields: Partial<T>) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields: Partial<T>) {
    const data = { ...fields };
    delete data?.id;
    delete data?.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    const isOk = result?.result?.ok ? true : false;
    const modifiedCount = result?.result?.nModified;
    return { ok: isOk, modifiedCount };
  }

  async function deleteOne(filter: Partial<T>) {
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    const isOk = result?.result?.ok ? true : false;
    return { ok: isOk };
  }

  return {
    findOne,
    findMany,
    createOne,
    updateOne,
    deleteOne
  };
}

export default createModel;
