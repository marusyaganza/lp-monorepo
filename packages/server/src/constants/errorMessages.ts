export const ERROR_MESSAGES = {
  NOT_AUTHENTICATED: 'Login to perform this operation',
  NOT_AUTHORIZED: 'You do not have permission to perform this operation',
  INCORRECT_CREDS: 'email or password is incorrect',
  SIGN_UP_FAILED: 'sign up operation failed',
  SING_UP_UNAVAILABLE:
    'Sign up operation is not available now. Sorry for inconvenience.',
  GAME_NOT_FOUND: 'Game not found',
  GAME_GENERATION_FAILED: 'Game data generation failed'
};
