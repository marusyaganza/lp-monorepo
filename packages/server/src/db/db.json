{
  "users": [
    {
      "email": "member@member.com",
      "password": "password",
      "role": "ADMIN",
      "verified": false,
      "avatar": "http",
      "createdAt": 1571213104370,
      "id": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "Test",
      "words": []
    },
    {
      "email": "me@test.com",
      "password": "123Test25",
      "name": "John Doe",
      "createdAt": 1652636040237,
      "id": "JG7MxQtfYykJQt9AB1lpZ",
      "role": "MEMBER"
    },
    {
      "email": "me2@test.com",
      "password": "123Test25",
      "name": "Jane Doe",
      "createdAt": 1652636259630,
      "id": "r039exmEjZRLk13s3GT62",
      "role": "MEMBER"
    }
  ],
  "words": [
    {
      "userId": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652607929132,
      "id": "DzOlEXTD39XQUnZELKXWG"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652608309387,
      "id": "BEqcqBjgV1D_i6zsqCIc7"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652615260601,
      "id": "rDlA84J5Mi0gppS42iIeM"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652615321617,
      "id": "18lsQ2U7oK572AIyitP2-"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652615498759,
      "id": "qTqmLBH9q-RMaQ1pqv0Pd"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652615597684,
      "id": "vlSe4J1KIBDW3jc5Ss5lN"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652615631503,
      "id": "n4p9HiUqfr8drzF7YySRD"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652615807151,
      "id": "MUS3RKn1HIrm0gz69UWR3"
    },
    {
      "user": "Z2hlvDmLOSRZGebC_Z_aA",
      "name": "egalitarian",
      "defs": [
        "a person who believes that everyone is equal and should have the same rights and opportunities"
      ],
      "particle": "noun",
      "examples": [
        "He described himself as ‘an egalitarian’."
      ],
      "createdAt": 1652634364376,
      "id": "ZNj_8cCpO2T6r-GT3HmGT"
    }
  ]
}