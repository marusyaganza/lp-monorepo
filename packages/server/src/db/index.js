const low = require('lowdb');
const FileSync = require('lowdb/adapters/FileSync');
const createModel = require('./models');

const adapter = new FileSync('src/db/db.json');
const db = low(adapter);

db.defaults({ words: [], users: [] });

module.exports = {
  models: {
    Word: createModel(db, 'words'),
    User: createModel(db, 'users')
  },
  db
};
