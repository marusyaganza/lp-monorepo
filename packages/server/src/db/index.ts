import { User, Word, SignUpInput, Maybe, AuthUser } from 'generated/graphql';

type UpdateResult = {
  ok: boolean;
  modifiedCount: number;
};

type DeleteResult = {
  ok: boolean;
};

type FindOneFunc<T> = (filter: Partial<T>) => Promise<T>;
type FindManyFunc<T> = (filter: Partial<T>) => Promise<T[]>;
type CreateOneFunc<T> = (filter: Partial<T>) => Promise<T>;
type UpdateOneFunc<T> = (filter: Partial<T>) => Promise<UpdateResult>;
type DeleteOneFunc<T> = (filter: Partial<T>) => Promise<DeleteResult>;

export type ModelType<T> = {
  findOne: FindOneFunc<T>;
  findMany: FindManyFunc<T>;
  createOne: CreateOneFunc<T>;
  updateOne: UpdateOneFunc<T>;
  deleteOne: DeleteOneFunc<T>;
};

// TODO: check if this type can be improved
export type UserModel = User & SignUpInput & AuthUser;

export type ModelsType = {
  Word: ModelType<Word>;
  User: ModelType<UserModel>;
};

export type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

export interface DbData extends ModelDataType {
  _id?: string;
}

export enum Table {
  WORD = 'words',
  USER = 'users'
}
