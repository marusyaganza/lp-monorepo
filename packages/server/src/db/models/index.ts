import { UserModel, UserModelType } from './User';
import { WordModel, WordModelType } from './Word';
import { WordTagModel, WordTagModelType } from './WordTag';

export interface ModelsType {
  User: UserModelType;
  Word: WordModelType;
  WordTag: WordTagModelType;
}

export const models: ModelsType = {
  User: UserModel,
  Word: WordModel,
  WordTag: WordTagModel
};
