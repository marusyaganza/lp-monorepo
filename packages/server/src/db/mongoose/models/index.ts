import { UserModel } from './User';
import { WordModel } from './Word';

export const models = {
  User: UserModel,
  Word: WordModel
};
