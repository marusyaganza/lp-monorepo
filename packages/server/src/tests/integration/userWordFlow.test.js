import {
  createTestServer,
  connectToDb,
  dropDb,
  seedDb,
  disconnectFromDb,
  getErrorMessageFromGQL,
  getDataFromGQL
} from '../helpers';
import bcrypt from 'bcryptjs';

import { models } from '../../db/models';
import { testData } from '../mocks/dbTestData';
import { wordsQuery, userQuery, wordByIdQuery } from '../mocks/gqlQueries';
import { generateGameData } from '../../generateGameData';
import { mutations } from '../mocks/gqlMutations';
import {
  createToken,
  getUserFromToken,
  hashPassword,
  validatePassword
} from '../../auth';
import { ERROR_MESSAGES } from '../../constants/errorMessages';

const userCreateInput = {
  email: 'test2@test.com',
  password: 'password',
  firstName: 'User',
  lastName: 'Test'
};

jest.mock('jsonwebtoken', () => {
  return {
    sign: jest.fn(() => `mock_token`),
    verify: jest.fn(() => 'user')
  };
});

const serverContext = {
  createToken,
  getUserFromToken,
  hashPassword,
  validatePassword,
  generateGameData,
  models
};

const { mutate } = createTestServer(serverContext);
const mockUser = { id: '6480560e8cad1841ed6b4011', role: 'MEMBER' };

const snapshotConfig = {
  id: expect.any(String)
};

/**
 * User can sign up
 * review his/her account infromation
 * review his/her words
 * add words
 * review his/her words
 * update a word
 * delete a word
 */

//TODO add test to check if unauthenticated user cannot perform auth operations
describe('User', () => {
  let addedWordId;
  let createdUser;

  beforeAll(async () => {
    await connectToDb();
    await dropDb();
    await seedDb();
  });

  afterAll(async () => {
    await disconnectFromDb();
    jest.clearAllMocks();
  });

  test('can sign up', async () => {
    const res = await mutate({
      mutation: mutations.signUpMutation,
      variables: {
        input: userCreateInput
      }
    });
    const { id, role } = getDataFromGQL(res).signUp;
    createdUser = { id, role };
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res).signUp).toMatchSnapshot(snapshotConfig);
  });

  test('can login with correct password', async () => {
    bcrypt.compare.mockReturnValueOnce(true);
    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: {
        input: {
          email: userCreateInput.email,
          password: userCreateInput.password
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res).login).toMatchSnapshot();
  });

  test('cannot login with incorrect password', async () => {
    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: { input: { email: 'test@test.com', password: 'password!' } }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.INCORRECT_CREDS);
  });

  test('cannot login with invalid email', async () => {
    const res = await mutate({
      mutation: mutations.loginMutation,
      variables: { input: { email: 'test2test.com', password: 'password' } }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.INCORRECT_CREDS);
  });

  test('cannot get the words list if user record in DB does not exist', async () => {
    const { query } = createTestServer({ ...serverContext, user: mockUser });
    const res = await query({ query: wordsQuery });
    expect(getErrorMessageFromGQL(res)).toEqual(
      ERROR_MESSAGES.NOT_AUTHENTICATED
    );
  });

  test('can get the words list', async () => {
    const { query } = createTestServer({ ...serverContext, user: createdUser });
    const res = await query({ query: wordsQuery });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res).wordsPerPage).toMatchSnapshot();
  });

  test('can get see the information about his/her account', async () => {
    const { query } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await query({ query: userQuery });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res)).toMatchSnapshot();
  });

  test('can save a word', async () => {
    const { mutate } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await mutate({
      mutation: mutations.saveWordMutation,
      variables: { input: testData.createWordInput }
    });
    addedWordId = getDataFromGQL(res).saveWord.id;
    expect(getDataFromGQL(res).saveWord).toMatchSnapshot(snapshotConfig);
  });

  test('nonexistent user cannot save a word', async () => {
    const { mutate } = createTestServer({
      ...serverContext,
      user: mockUser
    });
    const res = await mutate({
      mutation: mutations.saveWordMutation,
      variables: { input: testData.createWordInput }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.NOT_AUTHORIZED);
  });

  test('can save more words and view all of them', async () => {
    const { mutate, query } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    await mutate({
      mutation: mutations.saveWordMutation,
      variables: { input: testData.createWordInput2 }
    });
    const res = await query({ query: wordsQuery });
    const result = getDataFromGQL(res).wordsPerPage;
    expect(result.words.length).toBe(2);
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(result.hasNext).toEqual(false);
    expect(result.wordsCount).toBe(2);
    const words = result.words;
    const firstWordIndex = words.findIndex(w => w.id === addedWordId);
    expect(firstWordIndex).toBe(1);
  });

  test('can save more words and view all of them in reverse order', async () => {
    const { mutate, query } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    await mutate({
      mutation: mutations.saveWordMutation,
      variables: { input: testData.createWordInput2 }
    });
    const res = await query({
      query: wordsQuery,
      variables: { input: { isReverseOrder: true } }
    });
    expect(getDataFromGQL(res).wordsPerPage.words.length).toBe(2);
    const words = getDataFromGQL(res).wordsPerPage.words;
    const firstWordIndex = words.findIndex(w => w.id === addedWordId);
    expect(firstWordIndex).toBe(0);
  });

  test('cannot save a word with the same uuid', async () => {
    const { mutate } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await mutate({
      mutation: mutations.saveWordMutation,
      variables: { input: testData.createWordInput }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'word with uuid 74046e79-e4c9-4b52-ac96-cb7ae98fb601 is already added'
    );
  });

  test('can update a word', async () => {
    const { mutate } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await mutate({
      mutation: mutations.updateWordMutation,
      variables: { input: { ...testData.updateWordInput, id: addedWordId } }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res)).toMatchSnapshot();
  });

  test('cannot update immutable properties of a word', async () => {
    const { mutate } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await mutate({
      mutation: mutations.updateWordMutation,
      variables: { input: { ...testData.updateWordInput2, id: addedWordId } }
    });
    expect(getErrorMessageFromGQL(res)).toBeDefined();
  });

  test('can find a word by id', async () => {
    const { query } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await query({
      query: wordByIdQuery,
      variables: { wordId: addedWordId }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res).word).toMatchSnapshot(snapshotConfig);
  });

  test('can delete a word', async () => {
    const { mutate, query } = createTestServer({
      ...serverContext,
      user: createdUser
    });
    const res = await mutate({
      mutation: mutations.deleteWordMutation,
      variables: { deleteWordId: addedWordId }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(getDataFromGQL(res).deleteWord).toBe(
      `word with id ${addedWordId} was deleted`
    );
    const words = await query({
      query: wordsQuery,
      variables: { input: { isReverseOrder: true } }
    });
    expect(getDataFromGQL(words).wordsPerPage.words.length).toBe(1);
  });
});
