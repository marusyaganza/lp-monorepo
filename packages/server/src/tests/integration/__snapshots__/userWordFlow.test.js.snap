// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User can find a word by id 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "word": {
          "additionalInfo": "very useful information",
          "audioUrl": "mock.mp3",
          "defs": [
            {
              "def": "{bc} a strong substance that stretches and that is made out of chemicals or from the juice of a tropical tree",
              "examples": [
                "a rubber plantation",
                "a rubber tree",
              ],
            },
            {
              "def": "{b}{it}rubbers{/it}{/b} {it}US{/it}, {it}informal + old-fashioned{/it} {bc} shoes or boots that are made of rubber and that fit over your regular shoes to keep them dry",
              "examples": [],
            },
          ],
          "imgUrl": "mock.jpeg",
          "isOffensive": false,
          "level": "B2",
          "name": "egalitarian",
          "particle": "noun",
          "stems": [
            "egalitarian",
            "egalitarianism",
            "egalitarians",
          ],
          "transcription": "iˌɡælɪˈteəriən",
          "user": "6480560e8cad1841ed6b4011",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can get see the information about his/her account 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "user": {
          "email": "test@test.com",
          "firstName": "User",
          "lastName": "Test",
          "role": "MEMBER",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can get the words list 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "words": [],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can login with correct password 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "login": {
          "email": "test@test.com",
          "firstName": "User",
          "lastName": "Test",
          "token": "mock_token",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can save a word 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "saveWord": {
          "additionalInfo": "additional info",
          "audioUrl": "mock.mp3",
          "defs": [
            {
              "def": "a person who believes that everyone is equal and should have the same rights and opportunities",
              "examples": [
                "He described himself as ‘an {it}egalitarian{/it}’.",
              ],
            },
            {
              "def": "aiming for equal wealth, status, etc., for all people",
              "examples": [
                "He is a committed {it}egalitarian{/it}",
                "Even in the {it}egalitarian{/it} Nordic country, Marin felt hergender and age sometimes received too much emphasis. Kostya Manenkov And Karl Ritter, USA TODAY, 21 Aug. 2022",
              ],
            },
          ],
          "id": "mockId",
          "imgUrl": "mock.jpeg",
          "isOffensive": false,
          "level": "C1",
          "name": "egalitarian",
          "particle": "noun",
          "stems": [
            "egalitarian",
            "egalitarianism",
            "egalitarians",
          ],
          "tags": [
            {
              "color": "pink",
              "text": "academic",
            },
          ],
          "transcription": "iˌɡælɪˈteəriən",
          "user": "6480560e8cad1841ed6b4011",
          "uuid": "74046e79-e4c9-4b52-ac96-cb7ae98fb601",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can save more words and view all of them 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "words": [
          {
            "audioUrl": "mock.mp3",
            "defs": [
              {
                "def": "a person who believes that everyone is equal and should have the same rights and opportunities",
              },
              {
                "def": "aiming for equal wealth, status, etc., for all people",
              },
            ],
            "isOffensive": false,
            "level": "C1",
            "name": "egalitarian",
            "particle": "noun",
            "transcription": "iˌɡælɪˈteəriən",
          },
          {
            "audioUrl": null,
            "defs": [
              {
                "def": "{bc} a strong substance that stretches and that is made out of chemicals or from the juice of a tropical tree",
              },
              {
                "def": "{it}British{/it}",
              },
              {
                "def": "{b}{it}rubbers{/it}{/b} {it}US{/it}, {it}informal + old-fashioned{/it} {bc} shoes or boots that are made of rubber and that fit over your regular shoes to keep them dry",
              },
            ],
            "isOffensive": null,
            "level": null,
            "name": "rubber",
            "particle": "noun",
            "transcription": null,
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can sign up 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "signUp": {
          "email": "test@test.com",
          "firstName": "User",
          "lastName": "Test",
          "token": "mock_token",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User can update a word 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "updateWord": {
          "additionalInfo": "very useful information",
          "audioUrl": "mock.mp3",
          "defs": [
            {
              "def": "{bc} a strong substance that stretches and that is made out of chemicals or from the juice of a tropical tree",
              "examples": [
                "a rubber plantation",
                "a rubber tree",
              ],
            },
            {
              "def": "{b}{it}rubbers{/it}{/b} {it}US{/it}, {it}informal + old-fashioned{/it} {bc} shoes or boots that are made of rubber and that fit over your regular shoes to keep them dry",
              "examples": [],
            },
          ],
          "imgUrl": "mock.jpeg",
          "isOffensive": false,
          "level": "B2",
          "name": "egalitarian",
          "particle": "noun",
          "stems": [
            "egalitarian",
            "egalitarianism",
            "egalitarians",
          ],
          "tags": [
            {
              "color": "pink",
              "text": "academic",
            },
          ],
          "transcription": "iˌɡælɪˈteəriən",
          "user": "6480560e8cad1841ed6b4011",
          "uuid": "74046e79-e4c9-4b52-ac96-cb7ae98fb601",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User cannot login with incorrect password 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": null,
      "errors": [
        {
          "extensions": {
            "code": "INTERNAL_SERVER_ERROR",
          },
          "locations": [
            {
              "column": 5,
              "line": 3,
            },
          ],
          "message": "email or password is incorrect",
          "path": [
            "login",
          ],
        },
      ],
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User cannot login with invalid email 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": null,
      "errors": [
        {
          "extensions": {
            "code": "INTERNAL_SERVER_ERROR",
          },
          "locations": [
            {
              "column": 5,
              "line": 3,
            },
          ],
          "message": "email or password is incorrect",
          "path": [
            "login",
          ],
        },
      ],
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`User cannot save a word with the same uuid 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": null,
      "errors": [
        {
          "extensions": {
            "code": "INTERNAL_SERVER_ERROR",
          },
          "locations": [
            {
              "column": 3,
              "line": 2,
            },
          ],
          "message": "word with uuid 74046e79-e4c9-4b52-ac96-cb7ae98fb601 is already added",
          "path": [
            "saveWord",
          ],
        },
      ],
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
